util/winutil/svcdiag_windows.go

```diff
@@ -0,0 +1,303 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package winutil
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/mgr"
+	"tailscale.com/types/logger"
+)
+
+// LogSvcState obtains the state of the Windows service named rootSvcName and
+// all of its dependencies, and then emits that state to logf.
+func LogSvcState(logf logger.Logf, rootSvcName string) {
+	logEntries := []svcStateLogEntry{}
+
+	walkFn := func(svc *mgr.Service, config mgr.Config) {
+		status, err := svc.Query()
+		if err != nil {
+			logf("Failed retrieving Status for service %q: %v", svc.Name, err)
+		}
+
+		logEntries = append(logEntries, makeLogEntry(svc, status, config))
+	}
+
+	err := walkServices(rootSvcName, walkFn)
+	if err != nil {
+		logf("LogSvcState error: %v", err)
+		return
+	}
+
+	json, err := json.MarshalIndent(logEntries, "", "  ")
+	if err != nil {
+		logf("Error marshaling service log entries: %v", err)
+		return
+	}
+
+	var builder strings.Builder
+	builder.WriteString(`State of service "`)
+	builder.WriteString(rootSvcName)
+	builder.WriteString(`" and its dependencies:`)
```

Comment:
Nitpick: this doesn't escape; one could instead do:
```suggestion
	builder.WriteString(`State of service `)
	fmt.Fprintf(&builder, "%q", rootSvcName)
	builder.WriteString(` and its dependencies:`)
```